scripts/safe_template_checker.py

```diff
@@ -0,0 +1,498 @@
+
+    def __init__(self, full_path):
+        self.full_path = full_path
+        self.errors = []
+
+    def prepare_results(self, file_string):
+        line_breaks = _get_line_breaks(self, file_string)
+        for error in self.errors:
+            error.prepare_results(self.full_path, file_string, line_breaks)
+
+    def print_results(self, options):
+        if options['is_quiet']:
+            print self.full_path
+        else:
+            for error in self.errors:
+                error.print_results(options)
+
+
+class MakoTemplateChecker(object):
+
+    _skip_mako_dirs = _skip_dirs
+
+    _results = []
+
+    def process_file(self, directory, file_name):
+        """
+        Process file to determine if it is a Mako template file and
+        if it is safe.
+
+        Arguments:
+            directory (string): The directory of the file to be checked
+            file_name (string): A filename for a potential Mako file
+
+        Side effects:
+            Adds detailed results to internal data structure for
+            later reporting
+
+        """
+        if not self._is_mako_directory(directory):
+            return
+
+        # TODO: When safe-by-default is turned on at the platform level, will we:
+        # 1. Turn it on for .html only, or
+        # 2. Turn it on for all files, and have different rulesets that have
+        #    different rules of .xml, .html, .js, .txt Mako templates (e.g. use
+        #    the n filter to turn off h for some of these)?
+        # For now, we only check .html and .xml files
+        if not (file_name.lower().endswith('.html') or file_name.lower().endswith('.xml')):
+            return
+
+        self._load_and_check_mako_file_is_safe(directory + '/' + file_name)
+
+    def print_results(self, options):
+        for result in self._results:
+            result.print_results(options)
+
+    def _is_mako_directory(self, directory):
+        if _is_skip_dir(self._skip_mako_dirs, directory):
+            return False
+
+        if (directory.find('/templates/') >= 0) or directory.endswith('/templates'):
+            return True
+
+        return False
+
+    def _load_and_check_mako_file_is_safe(self, mako_file_full_path):
+        mako_template = _load_file(self, mako_file_full_path)
+        results = FileResults(mako_file_full_path)
+        self._check_mako_file_is_safe(mako_template, results)
+        if len(results.errors) > 0:
+            self._results.append(results)
+
+    def _check_mako_file_is_safe(self, mako_template, results):
+        has_page_default = self._has_page_default(mako_template, results)
+        if not has_page_default:
+            results.errors.append(BrokenRule(Rules.mako_missing_default))
+        self._check_mako_expressions(mako_template, has_page_default, results)
+        results.prepare_results(mako_template)
+
+    def _has_page_default(self, mako_template, results):
+        page_h_filter_regex = re.compile('<%page expression_filter=(?:"h"|\'h\')\s*/>')
```

Comment:
There's no advantage to compiling these before using them, and it forces you to come up with a name for them.